### DISCUSSION TITLE
Posterize and Threshold

### DISCUSSION THREAD

**Discussion by flowerhorse** [brunoherbelin/vimix#83] [Ideas]
At the moment posterizing switches from Fullcolor to 1 Color, (which means black) to 16C.
I think posterizing should start at least at 2color, so that the viewer is not suddenly in the dark. A black screen can be realized in another way.

My suggestion would be a slider left right operation with a range of 2 - 265 colors.
At the beginning OR AT THE END the fullrange values should be restored:
FR-2C-3C-4C...  ... 253C-254C-255C-256c-FR
So you could use the slider twice:
From posterizing 2C it would be hard cut, from 256C in the other direction it would creep in....

Also, the threshold slider doesn't seem to affect posterizing.
It can also be adjusted very very badly. It moves in much too large steps. Treshold is fineadjusting. A logarithmic slider would also be appropriate here.
The screen changes to different layers of black, but actually I should be able to set the threshold for the color separation with the Threshold to push the image a bit more...

Both sliders should be fine adjustable. Is it possible to build a slider that works logarithmic from its relative position, not from 'absolute' Position?

I would also like to be able to dynamically change the color of the blacks with 'Hue'.
At the moment 'Hue' only works in the range of reduced colors, which is much too little.
There the full color spectrum operation would have to be dynamically attainable there.

**Comment by brunoherbelin** [comment]
Thanks for your attention to details ! 
I also remember thinking that it was not satisfying that the posterization starts abruptly, and not from 'none' to 256 until 2 colors... but there were always other things to do in vimix, so never took the time... until now :)

Following some of your suggestions, I did the following to posterize and threshold sliders:
- value of posterize now range from 'Full range', to 255, down to 1 color.
- posterize slider is logarithmic, ie. fast around 255 and slower around 1
- threshold increment step is lowered to 0.001

> Also, the threshold slider doesn't seem to affect posterizing.

I am not sure I understand your comment : it should be the opposite IMHO : code is to first limit the number of colors, so that the threshold boundary is discretized by the steps of the posterization. You can try with e.g. posterize 3 colors and vary the threshold that jumps from 1 boundary to another.

> Both sliders should be fine adjustable. Is it possible to build a slider that works logarithmic from its relative position, not from 'absolute' Position?

humm, I guess I see what you mean but am not sure how to implement it... 
Challenge accepted :)

> I would also like to be able to dynamically change the color of the blacks with 'Hue'.
At the moment 'Hue' only works in the range of reduced colors, which is much too little.
There the full color spectrum operation would have to be dynamically attainable there.

Please give an example of what you mean.
On the computation side, by definition, Hue shift does not affect black and white as they are 0% and 100% lightness (or value)
(https://en.wikipedia.org/wiki/HSL_and_HSV)

**Comment by flowerhorse** [comment]
> I am not sure I understand your comment : it should be the opposite IMHO : code is to first limit the number of colors, so that the threshold boundary is discretized by the steps of the posterization. You can try with e.g. posterize 3 colors and vary the threshold that jumps from 1 boundary to another.

Yes, that's how I would have liked it.
With me, when I start it, the image turns black and white. I can make faces more recognizable in black and white, but the posterization colors are gone in all settings. When threshold is set to zero, the posterization colors are visible again. There would be 'teamwork' of the controllers nice. Nuanced elaboration of parts of the image is not possible with it
I tried your proposal with the same result.

> Please give an example of what you mean.
On the computation side, by definition, Hue shift does not affect black and white as they are 0% and 100% lightness (or value)
(https://en.wikipedia.org/wiki/HSL_and_HSV)

Following a test picture with the whole HUE range:
![Screenshot_20230413_015603](https://user-images.githubusercontent.com/125514250/231613331-55fe055c-d51e-4037-996b-670d88f952f4.png)


![Screenshot_20230413_020122](https://user-images.githubusercontent.com/125514250/231613463-0e0473fc-5f87-43e6-b8fe-7cc978df3058.png)
After Posterization 3: There are seven(!) colours now: (black, white, light grey, dark grey,) red, green, and brown. Something is already wrong here..
The HUE bar has changed to green, red,black. If I slide in HUE Hue has only three colours to manipulate the picture with. The position of the coloured areas changes then. 
I think the picture is internal converted now into a three(?) colour picture so Hue is limitated in it's colour range now to a three colour file and can give only these three colours according to the rule of posterisation.
If I'am right it would be better not to convert the picture, -to leave it at fullrange colour for the file- , but reduced to three colours within in this fullrange picture or excluding HUE from the rule of reducing it's full range to x-posterized colours.
In this case HUE could use the whole possible colour range within the three colour reduced picture and could change it to three colours which are determined mathematically from it depending on state of the Slider. In result this gave more colour possibilities while sliding in HUE and the effect wouldn't be as puristic as shown  above.

**Comment by flowerhorse** [comment]
Posterize works very well now. With the many color levels and the possibilities of gamma, brightness, contrast, saturation and hue you can now work very nuanced.

Threshold is also much better now, much easier to control than before, still a bit jerky visually, but that may be due to the contrast of the material.
The new position at the bottom of the menu is also better. There it is a bit isolated from the beginning because it mainly works solitary.
In principle one would need Threshold now more because over the setting Posterize 2 in connection with the contrast, brightness and gammaregeler also different Thresholds can be realized.
In my opinion, the changes have improved both the ease of use and the possibilities significantly.
Are you also satisfied with it?

**Comment by brunoherbelin** [comment]
Thanks for the extensive testing
Already good if the threshold and posterize sliders are more usable.

I'm not closing this discussion yet as I might check on the GLSL code for possible improvement.
